tools/approval-service/internal/service/ttleventcache.go

```diff
@@ -0,0 +1,120 @@
+/*
+ *  Copyright 2025 Gravitational, Inc
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package service
+
+import (
+	"sync"
+	"time"
+)
+
+// cacheEntry stores ID and expiry time for an item in the expiry list
+type cacheEntry struct {
+	id     string
+	expiry time.Time
+}
+
+// TTLEventCache is a thread-safe TTL-based event cache.
+// It uses "evict-on-write" semantics: expired entries are cleaned up during calls to TryAdd.
+//
+// The cache prevents re-processing of the string-based key for the configured time-to-live (ttl). While an entry
+// is in the cache, TryAdd will return false.
+type TTLEventCache struct {
+	mu         sync.Mutex
+	cache      map[string]time.Time // stores key -> expiry time for O(1) lookups
+	expiryList []cacheEntry         // stores (key, expiry time), sorted by expiry time
+	ttl        time.Duration
+}
+
+// NewTTLEventCache creates and starts a TTL-based in memory cache.
+// ttl specifies the time-to-live for each entry.
+// If ttl is zero or negative, a default of 15 seconds will be used.
+func NewTTLEventCache(ttl time.Duration) *TTLEventCache {
+	// set a logical default
+	if ttl <= 0 {
+		ttl = 15 * time.Second
+	}
+
+	c := &TTLEventCache{
+		cache: make(map[string]time.Time),
+		ttl:   ttl,
+	}
+
+	return c
+}
+
+// TryAdd attempts to add an entry to the cache.
+//
+// Returns true if the ID was added, or false if the ID is already in the cache and has not expired.
+//
+// Calls to TryAdd will prune any expired entries from the cache before performing the check.
+func (c *TTLEventCache) TryAdd(id string) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := time.Now()
+	c.evictExpired(now)
```

Comment:
This could potentially cause memory usage to balloon down the road under specific conditions if usage increased or changed. Can you add a comment here flagging this so that if we get hit by this in say three years, it's easier to identify what's going on?